x-pack/plugins/endpoint/server/routes/alerts.ts

```diff
@@ -5,27 +5,77 @@
  */
 
 import { IRouter } from 'kibana/server';
+import { SearchResponse } from 'elasticsearch';
+import { schema } from '@kbn/config-schema';
 
-import json from './sampledata.json';
+import { kibanaRequestToAlertListQuery } from '../services/endpoint/alert_query_builders';
+import { AlertData, AlertResultList } from '../../common/types';
+import { EndpointAppContext } from '../types';
 
-export function registerAlertRoutes(router: IRouter) {
-  router.get(
+export function registerAlertRoutes(router: IRouter, endpointAppContext: EndpointAppContext) {
+  router.post(
```

Comment:
why a post? seems more like a get?

Reply:
@kqualters-elastic It's what they used for the endpoints API, so I was keeping consistency. It allows for a request body, since not all clients will support a body with a GET request, and it's consistent with the ES APIs as well. I have no strong feelings about it though... I just want us to be consistent.

Reply:
Agreed with @kqualters-elastic, I think it's more RESTful

Reply:
I think the kibana/elastic way is to support post AND get

Reply:
We should support both, but GETs and POSTs have different use cases and shouldn't be treated as the same

Reply:
I meeeeean, I added a GET handler over the weekend... are you all looking at the latest? @oatkiller @alexk307 @oatkiller

Reply:
no, just commenting on this thread

Reply:
I still don't think the POST request should just return the list of alerts, should be only the GET

Reply:
@alexk307 Why not? It's the same way ES does it...

Reply:
https://en.wikipedia.org/wiki/Representational_state_transfer#Relationship_between_URI_and_HTTP_methods

I would argue it's not RESTful, and we should continue to try to observe that pattern even if Kibana as a whole doesn't. I think it's important to keep GETs idempotent. GETs should be limited to retrieving resources while POSTs should be limited to creating resources.

Reply:
@alexk307 GET is still idempotent in this case, which IMO is most important. Sometimes consistency is better than "correctness" and that's what I would argue in this case. I'll seek guidance from @oatkiller and @scunningham to make a call on this one... whatever is decided, we should be consistent at least across all of the Endpoint team's APIs.

Reply:
In addition, not all http client supports sending data in request body and that is why we used post. Because we want to be able to support all clients.

Reply:
Yep. For the GET request, I simplified to just take "page_size" and "page_index" as flat params in the query string. For POST, it requires the `paging_properties` nested structure in the BODY. @nnamdifrankie will we be flattening the POST body structure when the schema bug is fixed to align with what my GET request looks like?

Reply:
Are these APIs going to be used outside of the Kibana UI? Are users going to hit these APIs directly? If not, we only need to support the client we are using in Kibana.

Reply:
@alexk307 
![image](https://user-images.githubusercontent.com/35559/73680709-c30c0f00-468a-11ea-8a6a-921ced28eb4d.png)

Reply:
I guess I'm failing to figure out why we need to support `body`s in GET requests

Reply:
Well, we don't necessarily. Having a `body` gives us more flexibility (arbitrarily-nested JSON param structure), and that's something that the Elasticsearch APIs take advantage of for searching. When providing the same functionality as a GET, you have to be able to map all those params to flat querystring params (which is what I've done here with `page_size` and `page_index`). Having said that, given current requirements, there's nothing we can't support with simply a GET request. The primary argument that I have for providing POST support is maintaining consistency with Elasticsearch APIs (ES users will be accustomed to using POST requests like this). Also, this maintains consistency with what the Endpoint Management team has done so far (@paul-tavares / @nnamdifrankie).

And also, I guess I don't really understand what the big deal is, as long as GET requests are idempotent. That's the only requirement that, if not met, would result in confusion and unexpected behavior that is actually consequential IMO.

This discussion tbh seems a little like bikeshedding to me, and I'd like to focus our arguments on more consequential decisions. I totally get that it's not technically RESTful, but how much does that really matter? It obviously hasn't mattered for Elasticsearch, which has been a successful, developer-oriented project for many years.

Reply:
> This discussion tbh seems a little like bikeshedding to me, and I'd like to focus our arguments on more consequential decisions. I totally get that it's not technically RESTful, but how much does that really matter? It obviously hasn't mattered for Elasticsearch, which has been a successful, developer-oriented project for many years.

I think it's important to discuss this as we will be building and maintaining this application for years to come. We have the ability to define our own standards and style, so I think we should take the time to do so. IMHO I think we should stay away from having list/retrieval requests as POST requests. I'll let others comment, but that's just my 2c. Feel free to merge if you have consensus 👍

Reply:
Appreciate your input, @alexk307 ... didn't mean to imply that we shouldn't discuss it. I just don't personally think it's a hard decision to make, but perhaps my wish for consistency with Elasticsearch APIs is unreasonable. So I'm happy to get more input on this. I'd love to hear from @scunningham and @oatkiller , as they should be the ones who rule on things like this IMO.

Reply:
I agree with @kqualters-elastic and @alexk307  this seems a little odd if we're designing the API right now. 

@oatkiller `I think the kibana/elastic way is to support post AND get` - Is that documented anywhere? 

@madirey @nnamdifrankie Why are we trying to be consistent with Elasticsearch's APIs? We're adding functionality to Kibana not Elasticsearch - different APIs for different products should be okay.   Do either of you know why Elasticsearch implemented it that way?  Is there a reason to follow suite? 

@madirey the consequential decisions is supporting multiple APIs in various aspects (and adding the supporting testing). Does this API make sense for the Kibana app we're making?

Reply:
@james-elastic I think the origin of using GET and POST to do the same thing is https://www.elastic.co/guide/en/elasticsearch/guide/current/_empty_search.html#get_vs_post
I would think that ES users would be more accustomed to using GET requests since the POST is only provided as a workaround for clients that don't support GET with body.  Since we have a need for the body here, I favor dropping the POST handler and using GET without body only.

Reply:
FWIW I'm starting to tend to agree. Things could get muddled a bit when we start adding functionality that actually mutates data (whitelisting, archiving, etc). And if we're not concerned with aligning with ES, I'm totally fine switching to GET only.

Reply:
I'm in agreement that this type of API request should be a `GET`. If we get to a point where we need to support a `body`, then we can add the `POST` route at that time.  

The use case that comes to mind now in support of a `body` would be complex filters for a list - having a `body` where you can define those in JSON would probably simplify the implementation.

Reply:
It does look like some of the experimental documented saved objects APIs in Kibana use POST with a body as a way to retrieve information:
https://www.elastic.co/guide/en/kibana/current/saved-objects-api-bulk-get.html
https://www.elastic.co/guide/en/kibana/current/saved-objects-api-export.html

Kibana's general docs imply that POSTs should only be for adding things though: https://www.elastic.co/guide/en/kibana/current/using-api.html#api-calls

Reply:
So I guess I'll leave the POST support for now, but I'll implement GET within the app. If we need to switch, the support will be there, and if later we realize the post won't be needed, we can drop it. Sound good?